### ISSUE TITLE
It can not work when @NoArgsConstructor and @Builder and@Data added

### ISSUE THREAD

**Issue by yanjinbin** [rzwitserloot/lombok#1389]
just like this pic 
I found it will not work for applying constructor method something wrong 
hints: @Data has public constructor without args ,however , i need create a private no args constructor to make the function of  not instantiate sense ,
So, what should I do ?
![image](https://cloud.githubusercontent.com/assets/6295386/26140241/28563d72-3b08-11e7-80d2-8ebb9ecf9af9.png)

**Comment by Maaartinus** [comment]
Adding also `@AllArgsConstructor` should do.

**Comment by huehnerlady** [comment]
does that mean this is a bug? should the builder annotation work together with the NoArgsConstructor annotation?

**Comment by Maaartinus** [comment]
@huehnerlady I guess, so. `@Builder` sort of includes a "weak `@AllArgsConstructor`", which gets switched off by any explicit `@XArgsConstructor` annotation. As the `@Builder` can't work without the constructor, this should probably be strengthened:

It should *always* generate the needed constructor, unless a "colliding" constructor already exists (defined explicitly or using a Lombok annotation).

Here, "colliding" should mean that defining another one would make the compilation fail, but this isn't always possible, so "same number of arguments" gets used (as elsewhere in Lombok, too).

WDYT?

**Comment by zenglian** [comment]
@Maaartinus this does not work for \@SuperBuilder
Got the error when trying yaml serialize.

    compile group: 'com.esotericsoftware.yamlbeans', name: 'yamlbeans', version: '1.13'

**Comment by janrieke** [comment]
Great to hear.
But please do not delete your posts if someone directly replied to them, because it makes the discussion very hard to follow later on.

**Comment by wx8900** [comment]
@yanjinbin I add two annotations: @NoArgsConstructor and @AllArgsConstructor, it works. Good luck!

**Comment by ahmedhamdyse** [comment]
Thanks, it works for me

**Comment by bvn13** [comment]
> What is the exact error message you get? Is it a runtime error when you serialize, or a compile-time error?

I have an error

```
Error:(10, 1) java: constructor MdmAttribute in class com.bvn13.moex.mdm.buybackretriever.model.MdmAttribute cannot be applied to given types;
  required: no arguments
  found: java.lang.String,java.lang.String
  reason: actual and formal argument lists differ in length
```

My code is:

```java
@Builder
@Getter
@Setter
@NoArgsConstructor
@JacksonXmlRootElement(localName = "Attribute")
public class MdmAttribute {

    @JacksonXmlProperty(localName = "CodeAttr")
    private String codeAttr;

    @JacksonXmlProperty(localName = "ValueAttr")
    private String valueAttr;

    @Override
    public String toString() {
        return String.format("Attribute{CodeAttr=%s, ValueAttr=%s}", codeAttr, valueAttr);
    }

}
```

Also tried with @Data instead of @Getter + @Setter - causes the same error.

**Comment by mostafaghanyomar** [comment]
Hi @bvn13  I think it's the same as this issue too:
Duplicate of https://github.com/rzwitserloot/lombok/issues/816#
it all boils down to using `@Builder` in general whilst Using the default Constructor `new Foo()`, it doesn't matter if you added `@NoArgsConstructor` or not
The fastest and simplest solution is as @Maaartinus said, adding `@AllArgsConstructor` to your `@NoArgsConstructor`

**Comment by lgoldstein** [comment]
While adding `@AllArgsConstructor` is a **workaround** there are major software design issues with having such a constructor - especially for classes with many fields, and even more so when there are fields with same type. Imagine reading a code like this:
```java
Foo foo = new Foo(1, 2, true, false, 3, false, 5, true);
```
How can you tell what all these numbers and flags mean - what fields do they reference. This is where the builder pattern makes sense:
```java
Foo foo = Foo.builder()
    .minCount(1)
    .maxCount(2)
    .allowGaps(true)
    .allowOutOfRange(false)
   ...etc...
```
Why can't the `@Builder` annotation simply assume/enforce a `@NoArgsConstructor` and then simply invoke the relevant **setter** ?. If backward compatibility is an issue then let's invent a new annotation `@SetterBasedBuilder` that implies the above behavior.

**Comment by janrieke** [comment]
Because you may not have setters. And `final` fields cannot be set even directly.

The only way is to switch over to a constructor that uses the builder instance as its only parameter. However, this is backwards-incompatible and very likely to break existing code, so that's not an option.
If you want that approach, use `@SuperBuilder`.

Otherwise, use a `private` constructor. It's still ugly, but at least not very visible.

**Comment by antillean** [comment]
In [this comment](https://github.com/rzwitserloot/lombok/issues/1912#issuecomment-436062233) on #1912, @rzwitserloot said that this behaviour is intended.

**Comment by zhangyuhangk** [comment]
> 
> 
> Adding also `@AllArgsConstructor` should do.

Saved my ass. Thanks.

**Comment by cooligc** [comment]
This seems to be a critical defect . We cannot use `@Builder` on a JPA entity

**Comment by janrieke** [comment]
Please explain why adding `@AllArgsConstructor` does not solve this issue for you.

**Comment by cooligc** [comment]
> Please explain why adding `@AllArgsConstructor` does not solve this issue for you.

It does solve the issue after putting `@AllArgsConstructor`. Thanks for the help.
My point here was, Why to have unnecessary constructor in place which is not getting used.
Just for the shake of using `@Builder` we need to have `@NoArgsConstructor` and `@AllArgsConstructor`. @janrieke what is your thought on this ?

**Comment by Maaartinus** [comment]
@cooligc You're not stating it exactly. A lone `@Builder` works, as it creates its constructor itself. It's the presence of the `@NoArgsConstructor` which prevents it as an explicit `@XConstructor` annotation suppresses the implicit `@AllArgsConstructor`.

I'm not sure, if this should be changed. Obviously, the `@Builder` can't do anything with the `@NoArgsConstructor`, but the rules are already complicated.

Actually, I don't understand how `@Builder` and `@NoArgsConstructor` can make any sense together. With the latter I obviously want a  fully mutable object so why do I need the former?

**Comment by theangi** [comment]
Well, it may be useful in order to create cleaner tests with @Builder. I ended up with an object with @Builder, @NoArgsContructor and @AllArgsContructor

**Comment by nimo23** [comment]
We have to provide a no arg constructor and the full constructor needed by Builder. 

```
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class User {
..}
```

But why do I have to explicitly add `@AllArgsConstructor`? Lombok should automatically add the appropriate constructor if I annotate the class with `@Builder`, even if I also add `@NoArgsConstructor`.

Please enable this:

```
@Builder
@NoArgsConstructor
public class User {
..}
```

So the above

**Comment by sgupta304** [comment]
It should be noted that this is still a bug, even if it resolves the issue

**Comment by phisad** [comment]
> 
> 
> It should be noted that this is still a bug, even if it resolves the issue

It's not a bug, but a design choice to honor explicitly set annotations which totally makes sense.

**Comment by isen-ng** [comment]
I just ran into this and had to search the internet to find this issue.
I understand that this is a design choice, but by now, you should be able to tell that intuitively, people are going to put `@Builder @NoArgsConstructor` as their first choice and run into this exception.

Then they have to search the internet, read this thread and finally figure out that "by design" this exception is expected. 

Isn't this a hassle? Shouldn't software design be more intuitive?

I understand that this may be a design choice, but it seems that the UX is lacking. Perhaps, the error message could be more descriptive since we know this is a common pitfall. Eg, "You may have specified @Builder and @NoArgsContructor together. Try adding a @AllArksConstructor to make this go away".

**Comment by ivangreene** [comment]
Or better yet, just let the combination of `@Data` and `@Builder` be used as a signal that the constructor variants required by both need to be generated

**Comment by Art-B** [comment]
> > It should be noted that this is still a bug, even if it resolves the issue
> 
> It's not a bug, but a design choice to honor explicitly set annotations which totally makes sense.

Sorry but how is it not a bug when you need to specify both? I mean why do I need to add @AllArgsConstructor to get @NoArgsConstructor to behave? This at least violates the principle of least surprise

**Comment by phisad** [comment]
> > > It should be noted that this is still a bug, even if it resolves the issue
> > 
> > 
> > It's not a bug, but a design choice to honor explicitly set annotations which totally makes sense.
> 
> Sorry but how is it not a bug when you need to specify both? I mean why do I need to add @AllArgsConstructor to get @NoArgsConstructor to behave? This at least violates the principle of least surprise

If I recall correctly, you need `@AllArgsConstructor` to make `@Builder` to behave, which makes sense, because you explicitly forced a `@NoArgsConstructor` which overwrites the builders defaults. 

Well, as mentioned above, it could be forced that builder creates the all-args constructor regardless of other annotations which seems a good idea.

**Comment by Aleksander-D-92** [comment]
@Maaartinus 
"Actually, I don't understand how` @Builder` and `@NoArgsConstructor` can make any sense together."

A LOT of frameworks/libraries require the `@NoArgsConstructor` to work.
Like `hibernate` domain model classes
Like `RestTemplate` for API callls

So It's often just mandatory

**Comment by pcornelissen** [comment]
To elaborate on Aleksanders note:
Having a Builder inside your code doesn't mean that your frameworks will be able to use it. 

Having "only" an AllArgs constructor  with certain JSON Libs leads to pain. Hibernate as well. Other transport formats also lead to stress. There you still need setters and the NoArgConstructor.

Here is what I expected to happen when I have a class that has maybe the @Data annotation, when I add the @Builder annotation:
- Builder methods are created
- AllArgsConstructor for Builder is created (although I don't care for that, as my focus is on the Builder)
- Builder works
- The rest stays the same

But what happens is:
- Builder methods are created
- AllArgsConstructor is created
- Builder works
- Due to the fact that AllArgs is added, NoArgs is omitted

That is just something you don't expect and it requires deeper knowledge about how Lombok works. That is bad.

When I see @Builder I don'e extpect a negative effect on unrelated constructors. It just breaks the principle of least surprise. When you see @Builder(skipNoArgsConstructor=true) then I know what happens. No surprise here.

**Comment by Art-B** [comment]
> So It's often just mandatory. And we also some times want to use the @Builder to manually create our own objects from this classes.

Yes, often -for us- the builder is only used in tests and the NoArgsConstructor only by the libraries reflexively.

**Comment by mxmlnglt** [comment]
Any way this "issue" can be re-opened?
Because OP @yanjinbin "closed this on Nov 5, 2019" but without any comment to the matter, and given the recent comments, it seems the discussion is NOT over...

**Comment by pcornelissen** [comment]
It is for them it seems :-/

**Comment by pedroluiznogueira** [comment]
Is there anyway this can be reopened ?

**Comment by ivangreene** [comment]
@rzwitserloot here is some additional argumentation for providing different behavior when another `@XArgsConstructor` (where `X` is not `All`) and `@Builder` are combined:
* Currently, combining `@XArgsConstructor` and `@Builder` generates code which doesn't compile. This isn't a useful scenario for anyone
* If a user wanted to specifically avoid generating an all-args constructor when `@Builder` was present (I could see this happening if the user overrides the build() method, for example), this could still be expressed with `@AllArgsConstructor(access=AccessLevel.NONE)`
* It's a common pattern to use a Builder for all directly created instances of a bean, alongside using a deserializer which requires a no-args constructor and reflectively sets the fields of the bean. Thus it is common to have a bean with annotations like `@Getter @EqualsAndHashCode @Builder`, encounter an exception from the deserializer mentioning needing a no-args constructor, expect everything to work after adding `@NoArgsConstructor`, but find this has broken the basic functionality of Builder, which then has a non-obvious solution. Why is this a better outcome than retaining an all-args constructor?

I'm positing that retaining an all-args constructor in the presence of both `@Builder` and `@XArgsConstructor` serves a larger number of users than the current behavior does. The current behavior only serves those who need to **not** have an all args constructor, and have overridden the `build()` method on the builder

**Comment by rzwitserloot** [comment]
Your arguments make no sense. They all boil down to: "... and if that is not what you intended to do, simply add this simple annotation and that fixes the problem". But that's __already__ what we have now. Just add `@AllArgsConstructor` next to your `@Builder` annotation.

I'll help you out, then: The direction you want to take your argument is that having `@Builder` presume the `@AllArgsConstructor` is inherently less confusing than not doing so. Except, that argument is almost impossible to prove, and we disagree with it, so in the end this is a dead end if your aim is to convince us.

Whatever weirdo beanspec layout that requires you to hang lombok annotations on your class like its a christmas tree? We have other, better solutions for this, than messing with the definition of `@Builder`.

* The meta annotations idea (not implemented yet, but rawi has plans, and it's gotten the green light as far as feature idea goes): You make your own annotation that implies that whole christmas tree, and then you just annotate your types with that. You can then make one annotation that implies `@NoArgsConstructor @AllArgsConstructor @Builder` and call it `@BeanBuilder` or whatnot.

* We have an `extern` package which we feel is extremely underused. We tend not to come up with good ideas because we @rspilker and I both lean towards less dependencies. Especially dependencies that have a significant impact on how you structure your java code (Example: We both tend to steer away from JPA). However, we don't begrudge your use of JPA, spring/dagger/guice, Jackson/GSON, etcetera. We just don't accept any attempt to force 'play well with [external library] specifically' into a general lombok feature, which is why `extern` exists. If there is a specific library that effectively dictates that all classes you use with it have a highly specific structure (such as: Getters, builder, but also a no-args constructor and package-private setters, and let's say that you tend to always have some ID field that should just always exist, and be excluded from equals and hashCode impls), then prove it (as we won't be familiar with this library most likely) and we can then make an annotation _specifically_ for this library that does it all, with a single annotation.

**Comment by owenammann** [comment]
Chiming in here--since having a no args constructor and a builder is not mutually exclusive, you should be able to enable both of them without explicitly having an all args constructor (implicitly made by the builder). The reason I disagree with your disagreement is I needed to look it up and find this thread... But in clearer terms--the builder tag makes a builder, the no args constructor makes a no args constructor. If the no args constructor deletes the all args made by the builder, then the **builder breaks the builder design contract**. Therefore, a builder is not correctly made by the builder tag. To me, fixing this is not messing with the definition of builder at all, in fact it is preserving it.

**Comment by minervaaniket** [comment]
@AllArgsConstructor required if Builder is used with @NoArgsConstructor - because builder needs @AllArgsConstructor. - Simple Fix

if you only use @Builder - it will work fine because private parameterized constructor will be created by @Builder, but if you use any XArgsConstructor, it will override the builder constructor.
 
So if you use, @Builder and @NoArgsConstructor, then NoArgsConstructor will override the Builder's parameterized constructor. That is the reason you will get error like : 
  required: no arguments
  found: java.lang.String,java.lang.String
  reason: actual and formal argument lists differ in length

Because your @NoArgsConstructor was expecting no arguements, but builder is trying to use same by passing parameters.
